### DISCUSSION TITLE
Make Meteor.startup compatible with async functions

### DISCUSSION THREAD

**Discussion by perbergland** [meteor/meteor#12468] [Feature Requests]
Meteor.startup currently does not allow for async functions and in order to make it work, the Profile package also needs to be async-enabled, both for Profile.run and Profile.time.

I just checked in the release-3.0 branch and this has not happened there yet so there is nothing to backport to 2.x.

```boot.js
var callStartupHooks = Profile("Call Meteor.startup hooks", function () {
  // run the user startup hooks.  other calls to startup() during this can still
  // add hooks to the end.
  while (__meteor_bootstrap__.startupHooks.length) {
    var hook = __meteor_bootstrap__.startupHooks.shift();
    Profile.time(hook.stack || "(unknown)", hook);
  }
  // Setting this to null tells Meteor.startup to call hooks immediately.
  __meteor_bootstrap__.startupHooks = null;
});
```

This would need to become

```boot.js
var callStartupHooks = Profile("Call Meteor.startup hooks", async function () {
  // run the user startup hooks.  other calls to startup() during this can still
  // add hooks to the end.
  while (__meteor_bootstrap__.startupHooks.length) {
    var hook = __meteor_bootstrap__.startupHooks.shift();
    await Profile.time(hook.stack || "(unknown)", hook);
  }
  // Setting this to null tells Meteor.startup to call hooks immediately.
  __meteor_bootstrap__.startupHooks = null;
});
```

and consequently (release-3-0 branch)

```
async function startServerProcess() {
  const { AsyncLocalStorage } = require('async_hooks');
  global.asyncLocalStorage = new AsyncLocalStorage();

  await Profile.run('Server startup', async function() {
    // TODO[FIBERS] the if around loadServerBundles should be enough
    if (IS_FIBERS_ENABLED) {
      loadServerBundles();
      await callStartupHooks();
      runMain();
    } else {
      await global.asyncLocalStorage.run({ level: 'boot' }, async () => {
        loadServerBundles();
        await callStartupHooks();
        runMain();
      });
    }
  });
}
```

To be able to port packages like percolate:migrations, this work needs to happen fairly soon. Could you add it to the fibers plan?

**Comment by perbergland** [comment]
In 2.x it could be enough to call

```javascript
Promise.await(callStartupHooks())
```

as long as callStartupHooks is made async all the way

**Comment by filipenevola** [comment]
Hi @perbergland, we have a few branches in the works for 3.0.

This part calling startup hooks is still in the `release-3.0-tools` branch, you can see the code [here](https://github.com/meteor/meteor/blob/f7c419a3d96c0639e52116145590a24e0bedb4a6/tools/static-assets/server/boot.js#L441-L501) already with `async`.

We are working to unify all the branches for 3.0 as soon as possible but they have blockers between them so we are keeping multiple branches in order to move forward in parallel.

**Comment by perbergland** [comment]
Thanks @filipenevola!

Would love to see this backported to 2.x since it doesn’t break any APIs but is necessary for writing code that is compatible with 3.0
